packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -1868,6 +1871,212 @@ function commitPassiveLifeCycles(finishedWork: Fiber): void {
+      while (fiber !== null) {
+        if (fiber.child !== null) {
+          // Should we add a separate subtree tag for this?
+          const primarySubtreeTag = fiber.subtreeTag & LayoutSubtreeTag;
+          if (primarySubtreeTag !== NoSubtreeTag) {
+            invokeLayoutEffectsUnmountInDEV(fiber.child);
+          }
+        }
+
+        const effectTag = fiber.effectTag;
+        const current = fiber.alternate;
+        // This is a mount
+        if (current === null) {
+          if (effectTag & Update) {
+            switch (fiber.tag) {
+              case FunctionComponent:
+              case ForwardRef:
+              case SimpleMemoComponent:
+              case Block: {
+                invokeGuardedCallback(
+                  null,
+                  commitHookEffectListUnmount,
+                  null,
+                  HookLayout | HookHasEffect,
+                  fiber,
+                );
+                if (hasCaughtError()) {
+                  const unmountError = clearCaughtError();
+                  captureCommitPhaseError(fiber, unmountError);
+                }
+                break;
+              }
+              case ClassComponent: {
+                const instance = fiber.stateNode;
+                if (typeof instance.componentWillUnmount === 'function') {
+                  safelyCallComponentWillUnmount(fiber, instance);
+                }
+                break;
+              }
+            }
+          }
+        }
+        fiber = fiber.sibling;
+      }
+    }
+  }
+}
+
+function invokeLayoutEffectsMountInDEV(firstChild) {
+  if (enableDoubleInvokingEffects) {
+    // mount layout effects
+    if (__DEV__) {
+      let fiber = firstChild;
+      while (fiber !== null) {
+        if (fiber.child !== null) {
+          // Should we add a separate subtree tag for this?
+          const primarySubtreeTag = fiber.subtreeTag & LayoutSubtreeTag;
+          if (primarySubtreeTag !== NoSubtreeTag) {
+            invokeLayoutEffectsMountInDEV(fiber.child);
+          }
+        }
+
+        const effectTag = fiber.effectTag;
+        const current = fiber.alternate;
+        if (current === null) {
+          if (effectTag & Update) {
+            switch (fiber.tag) {
+              case FunctionComponent:
+              case ForwardRef:
+              case SimpleMemoComponent:
+              case Block: {
+                invokeGuardedCallback(
+                  null,
+                  commitHookEffectListMount,
+                  null,
+                  HookLayout | HookHasEffect,
+                  fiber,
+                );
+                if (hasCaughtError()) {
+                  const mountError = clearCaughtError();
+                  captureCommitPhaseError(fiber, mountError);
+                }
+                break;
+              }
+              case ClassComponent: {
+                const instance = fiber.stateNode;
+                instance.componentDidMount();
```

Comment:
Should we be wrapping this call with `invokeGuardedCallback` and checking `hasCaughtError` like we do with the others? (If not, why not?)

Reply:
So there's already an inconsistency between unmount and mount.

https://github.com/facebook/react/pull/19605 also surfaced that there's an inconsistency within them.

Mount uses a different technique. The workloop has a try/catch around it. https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberWorkLoop.new.js#L2651

Concretely this means that, for mounts, if there are multiple useEffect hooks, and one throws, the remaining ones in the same component won't get called. That's a bit different than unmounts and there are arguments to be made that we should change this behavior.

However, the issue in this case is that it does neither. So if something throws, all the remaining fibers will be ignored. Not just the remaining hooks.

A quick fix is probably just to use invokeGuardedCallback like Brian proposed.